File: frontend/src/api/publicBoard.ts
import { useMutation, useQuery } from "@tanstack/react-query";

import { api } from "./client";
import type { JobType, Seniority, WorkMode } from "./jobs";

export interface PublicJobSummary {
  id: string;
  slug: string;
  title: string;
  overview: string | null;
  applicant_count: number;
  work_mode: WorkMode | null;
  location: string | null;
  seniority: Seniority | null;
  job_type: JobType | null;
  salary_min: number | null;
  salary_max: number | null;
  salary_currency: string | null;
  created_at: string;
  org_name: string | null;
  org_logo_url: string | null;
  board_path: string | null;
}

export interface PublicBoardResponse {
  org_name: string;
  org_logo_url: string | null;
  org_description: string | null;
  org_office_location: string | null;
  org_contact_email: string | null;
  jobs: PublicJobSummary[];
}

export interface PublicScreeningQuestion {
  id: string;
  question_text: string;
  question_type: "text" | "number" | "boolean";
  required: boolean;
  position: number;
}

export interface PublicJobDetail {
  id: string;
  title: string;
  overview: string | null;
  description: string | null;
  is_technical_role: boolean;
  applicant_count: number;
  work_mode: WorkMode | null;
  location: string | null;
  seniority: Seniority | null;
  job_type: JobType | null;
  salary_min: number | null;
  salary_max: number | null;
  salary_currency: string | null;
  screening_questions: PublicScreeningQuestion[];
  posted_by_name: string;
  org_name: string | null;
  org_logo_url: string | null;
  created_at: string;
  board_path: string;
}

export function useOrgBoard(slug: string) {
  return useQuery({
    queryKey: ["public-board", slug],
    queryFn: async () => (await api.get<PublicBoardResponse>(`/public/boards/${slug}`)).data,
    enabled: !!slug,
    retry: false,
  });
}

export function useAllJobsBoard() {
  return useQuery({
    queryKey: ["public-board", "all"],
    queryFn: async () => (await api.get<PublicBoardResponse>("/public/boards/all")).data,
  });
}

export function usePublicJob(jobSlug: string) {
  return useQuery({
    queryKey: ["public-job", jobSlug],
    queryFn: async () => (await api.get<PublicJobDetail>(`/public/jobs/${jobSlug}`)).data,
    enabled: !!jobSlug,
    retry: false,
  });
}

export interface ApplyInput {
  full_name: string;
  email: string;
  phone: string;
  cover_letter?: string;
  years_of_experience: string;
  linkedin_url: string;
  github_url?: string;
  portfolio_url?: string;
  open_to_other_roles: boolean;
  answers: { question_id: string; answer: string }[];
  cv: File;
}

export function useApplyToJob(jobSlug: string) {
  return useMutation({
    mutationFn: async (input: ApplyInput) => {
      const form = new FormData();
      form.append("full_name", input.full_name);
      form.append("email", input.email);
      form.append("phone", input.phone);
      if (input.cover_letter) form.append("cover_letter", input.cover_letter);
      form.append("years_of_experience", input.years_of_experience);
      form.append("linkedin_url", input.linkedin_url);
      if (input.github_url) form.append("github_url", input.github_url);
      if (input.portfolio_url) form.append("portfolio_url", input.portfolio_url);
      form.append("open_to_other_roles", String(input.open_to_other_roles));
      form.append("answers_json", JSON.stringify(input.answers));
      form.append("cv", input.cv);
      const { data } = await api.post<{ eligible: boolean; message: string }>(
        `/public/jobs/${jobSlug}/apply`,
        form,
        { headers: { "Content-Type": "multipart/form-data" } },
      );
      return data;
    },
  });
}

File: backend/app/schemas/public_board.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class PublicJobSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    slug: str
    title: str
    overview: str | None
    applicant_count: int
    work_mode: str | None
    location: str | None
    seniority: str | None
    job_type: str | None
    salary_min: int | None
    salary_max: int | None
    salary_currency: str | None
    created_at: datetime
    # Only populated on the all-jobs board (org-specific boards already
    # know their own org from the top-level org_name) — see docs/10.
    org_name: str | None = None
    org_logo_url: str | None = None
    # Always populated (unlike org_name/org_logo_url above) — lets the
    # org name/logo on a job card link back to that org's own board even
    # from the all-jobs board. /jobs for the Freelance Org, /jobs/{slug}
    # for an Org tenant. Same value PublicJobDetail.board_path carries.
    board_path: str | None = None


class PublicBoardResponse(BaseModel):
    org_name: str
    # Org profile fields — null for the Freelance Org, which has no
    # org profile to show. See app/api/routers/org.py.
    org_logo_url: str | None = None
    org_description: str | None = None
    org_office_location: str | None = None
    org_contact_email: str | None = None
    jobs: list[PublicJobSummary]


class PublicScreeningQuestionOut(BaseModel):
    # Deliberately no expected_answer/min_value — the actual pass
    # criteria never ships to the public form. question_type/required
    # only drive how the form renders the input (text vs number field,
    # asterisk for required) and client-side validation, not the real
    # eligibility check (that's always re-evaluated server-side).
    id: uuid.UUID
    question_text: str
    question_type: str
    required: bool
    position: int


class PublicJobDetail(BaseModel):
    id: uuid.UUID
    title: str
    overview: str | None
    description: str | None
    is_technical_role: bool
    applicant_count: int
    work_mode: str | None
    location: str | None
    seniority: str | None
    job_type: str | None
    salary_min: int | None
    salary_max: int | None
    salary_currency: str | None
    screening_questions: list[PublicScreeningQuestionOut]
    # Who's hiring — the recruiter who owns the job, and (for an Org
    # tenant, not the Freelance Org) that org's name/logo. See docs/10.
    posted_by_name: str
    org_name: str | None
    org_logo_url: str | None
    created_at: datetime
    # Frontend route back to the board this job came from — /jobs for
    # the all-jobs board, /jobs/{slug} for an Org tenant. Lets the
    # post-apply "browse other jobs" link go back to the right board
    # instead of guessing.
    board_path: str


class ApplyResponse(BaseModel):
    eligible: bool
    message: str
